### PR TITLE
 feat: 杜绝浪费，在Job上增加相关特性，避免不必要的开销 #970

### PR DESCRIPTION
**Pull request by huangjiao-heart**

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/bk-ci?pullRequest=2454) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/bk-ci?pullRequest=2454) before we can accept your contribution.<br/><hr/>**johuang** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/bk-ci?pullRequest=2454) it.</sub>

File: src/backend/ci/core/dispatch/biz-dispatch/src/main/kotlin/com/tencent/devops/dispatch/service/PipelineDispatchService.kt [hunk 1]
```diff
@@ -109,7 +113,17 @@ class PipelineDispatchService @Autowired constructor(
...
         getDispatchers().forEach {
             if (it.canDispatch(pipelineAgentStartupEvent)) {
+                // JOB配额判断
+                if (!jobQuotaBusinessService.checkJobQuota(pipelineAgentStartupEvent, buildLogPrinter)) {
+                    it.onFailBuild(client, buildLogPrinter, pipelineAgentStartupEvent, ErrorType.USER, ErrorCodeEnum.JOB_QUOTA_EXCESS.errorCode, "系统JOB配额超限")
+                    return
+                }
                 it.startUp(pipelineAgentStartupEvent)
+                // 到这里说明JOB已经启动成功(但是不代表Agent启动成功)，开始累加使用额度
+                val vmType = JobQuotaVmType.parse(pipelineAgentStartupEvent.dispatchType)
```
Comment (irwinsun) [inline src/backend/ci/core/dispatch/biz-dispatch/src/main/kotlin/com/tencent/devops/dispatch/service/PipelineDispatchService.kt:123]: 这里要处理一下，如果dispatchType因为异常的值，导致parse异常抛出了。后续流程会都断掉了

File: src/backend/ci/core/dispatch/biz-dispatch/src/main/kotlin/com/tencent/devops/dispatch/dao/JobQuotaSystemDao.kt [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    fun list(dslContext: DSLContext): Result<TDispatchQuotaSystemRecord?> {
+        with(TDispatchQuotaSystem.T_DISPATCH_QUOTA_SYSTEM) {
+            return dslContext.selectFrom(this)
+                .fetch()
+        }
+    }
+
+    fun convert(record: TDispatchQuotaSystemRecord): JobQuotaSystem {
+        return JobQuotaSystem(
+            vmType = JobQuotaVmType.parse(record.vmType),
```
Comment (irwinsun) [inline src/backend/ci/core/dispatch/biz-dispatch/src/main/kotlin/com/tencent/devops/dispatch/dao/JobQuotaSystemDao.kt:59]: JobQuotaVmType.parse 可能会抛出异常，导致调用流程中断

File: src/backend/ci/core/dispatch/api-dispatch/src/main/kotlin/com/tencent/devops/dispatch/api/ServiceJobQuotaBusinessResource.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+@Api(tags = ["BUILD_JOBS_PROJECT_QUOTA"], description = "Job配额管理")
+@Path("/service/jobs/running")
+@Produces(MediaType.APPLICATION_JSON)
+@Consumes(MediaType.APPLICATION_JSON)
+interface ServiceJobQuotaBusinessResource {
+
+    @ApiOperation("上报一个JOB启动")
+    @POST
+    @Path("/{projectId}/{vmType}/{buildId}/{vmSeqId}")
+    fun addRunningJob(
```
Comment (irwinsun) [inline src/backend/ci/core/dispatch/api-dispatch/src/main/kotlin/com/tencent/devops/dispatch/api/ServiceJobQuotaBusinessResource.kt:53]: 同为POST方法的
@Path("/{projectId}/{vmType}/{buildId}/{vmSeqId}")路径
与
addRunningAgent 的 @Path("/agent/{projectId}/{buildId}/{vmSeqId}")

可能存在冲突 比如请求的projectId为agent时，会出现二义性，有可能会调用出错

Comment (huangjiao-heart) [reply src/backend/ci/core/dispatch/api-dispatch/src/main/kotlin/com/tencent/devops/dispatch/api/ServiceJobQuotaBusinessResource.kt:53]: 已修改

File: src/backend/ci/core/dispatch/biz-dispatch/src/main/kotlin/com/tencent/devops/dispatch/service/PipelineDispatchService.kt (hunk 1 above)
Comment (huangjiao-heart) [reply src/backend/ci/core/dispatch/biz-dispatch/src/main/kotlin/com/tencent/devops/dispatch/service/PipelineDispatchService.kt:123]: 已修改

**Review by irwinsun — APPROVED** [review]

File: src/backend/ci/core/common/common-auth/common-auth-blueking/src/main/kotlin/com/tencent/devops/common/auth/code/BkAuthServiceCode.kt [hunk 4]
```diff
@@ -26,10 +26,10 @@
 
 package com.tencent.devops.common.auth.code
 
-const val GLOBAL_SCOPE_TYPE = "system"
+// const val GLOBAL_SCOPE_TYPE = "system"
```
Comment (irwinsun) [inline src/backend/ci/core/common/common-auth/common-auth-blueking/src/main/kotlin/com/tencent/devops/common/auth/code/BkAuthServiceCode.kt:29]: 参数在别的地方有引用，不能注释掉

File: src/backend/ci/core/common/common-auth/common-auth-blueking/src/main/kotlin/com/tencent/devops/common/auth/code/BkAuthServiceCode.kt [hunk 5]
```diff
@@ -26,10 +26,10 @@
...
 const val PROJECT_SCOPE_TYPE = "project"
 // 蓝鲸持续集成平台产品ID
-const val BK_DEVOPS_SCOPE = "bk_ci"
+// const val BK_DEVOPS_SCOPE = "bk_ci"
```
Comment (irwinsun) [inline src/backend/ci/core/common/common-auth/common-auth-blueking/src/main/kotlin/com/tencent/devops/common/auth/code/BkAuthServiceCode.kt:32]: 参数在下面有引用，不能注释掉

File: src/backend/ci/core/project/api-project/build.gradle [hunk 6]
```diff
@@ -28,6 +28,7 @@ dependencies {
     compile project(":core:common:common-api")
     compile project(":core:common:common-web")
     compile project(":core:common:common-event")
+    compile project(":core:common:common-auth:common-auth-blueking")
```
Comment (irwinsun) [inline src/backend/ci/core/project/api-project/build.gradle:31]: 这里需要改为依赖 common-auth-api
blueking为具体业务实现，不应该作为api层的依赖

**Review by irwinsun — CHANGES_REQUESTED** [review]
需要针对上述进行必要修改

**Comment by irwinsun** [comment]
#970

File: src/backend/ci/core/project/api-project/build.gradle (hunk 6 above)
Comment (huangjiao-heart) [reply src/backend/ci/core/project/api-project/build.gradle]: 已调整

File: src/backend/ci/core/common/common-auth/common-auth-blueking/src/main/kotlin/com/tencent/devops/common/auth/code/BkAuthServiceCode.kt (hunk 5 above)
Comment (irwinsun) [reply src/backend/ci/core/common/common-auth/common-auth-blueking/src/main/kotlin/com/tencent/devops/common/auth/code/BkAuthServiceCode.kt:32]: 已经在AuthConstants上定义，此处重复，可移除

File: src/backend/ci/core/common/common-auth/common-auth-blueking/src/main/kotlin/com/tencent/devops/common/auth/code/BkAuthServiceCode.kt (hunk 4 above)
Comment (irwinsun) [reply src/backend/ci/core/common/common-auth/common-auth-blueking/src/main/kotlin/com/tencent/devops/common/auth/code/BkAuthServiceCode.kt:29]: 已经在AuthConstants上定义，此处重复，可移除

**Review by irwinsun — APPROVED** [review]

**Pull request merged**